subcommand/get-consul-client-ca/command_test.go

```diff
@@ -0,0 +1,388 @@
+package getconsulclientca
+
+import (
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"github.com/hashicorp/go-discover"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/freeport"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	"github.com/hashicorp/consul/tlsutil"
+	"github.com/mitchellh/cli"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRun_FlagsValidation(t *testing.T) {
+	t.Parallel()
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI: ui,
+	}
+
+	exitCode := cmd.Run([]string{
+		"-output-file", "",
+	})
+	require.Equal(t, 1, exitCode)
+	require.Contains(t, ui.ErrorWriter.String(), "-output-file must be set")
+}
+
+// Test that in the happy case scenario
+// we retrieve the CA from Consul and
+// write it to a file
+func TestRun(t *testing.T) {
+	t.Parallel()
+	outputFile, err := ioutil.TempFile("", "ca")
```

Comment:
Yes, the file will be created if it doesn't exist (under the hood, golang will make a system call that will create the file if it doesn't exist).

I went with the file because I like that you can simply write a string to it by calling `file.WriteString`. But I don't think it matters for the command. Curious why think dir is better.

Reply:
I was thinking that you want to test that the command writes a file to disk. When the command runs normally the file it's writing to won't already exist but in this test you're pre-creating the file.

With a temporary dir you can test that it properly writes a file to that dir and then delete the temporary dir.

Reply:
fwiw pre-creating a file when we want to test that it get's created doesn't make much sense to me. You're not calling `file.write` in this test.

Reply:
To clarify, the intention here is to test that the expected contents are written to the file, not that the file gets created if it doesn't exist. The command should work the same in both cases. If the file already exists, it should write over the contents. If the file doesn't exist, it will create it and write to it. Testing those two cases exhaustively doesn't make much sense to me though because it's equivalent to testing golang's ioutil library. I think it's sufficient to trust the API's promise that it will create the file if it doesn't exist and just check that the file contents are what we expect.